Fix proto resource 153 marshalling for autoupdate_* resources

hugoShaka opened the pull request:
This PR adds `types.Resource153ToLegacyV2()` which acts exactly like `types.Resource153ToLegacy()` but detects if the underlying resource153 implementation is a proto struct and uses `protojson` instead of regular json marshaling.

This fixes improper marshaling for resources `autoupdate_config`, `autoupdate_version`, and `autoupdate_agent_rollout`.

This is a band-aid solution, the proper way to address this would be to have the proto structs implement `json.Marshaler`.

This PR does not change `types.Resource153ToLegacy()` because:
- I have no idea what implications this will have for other resources, this might change how the resource is written on disk, exported audit logs, ...
- Automatic updates will need to be backported, we cannot retroactively change how teleport marshals resources.

Goal (internal): https://github.com/gravitational/cloud/issues/10289

hugoShaka committed:
Fix proto resource 153 marshalling

tigrato commented on api/types/resource_153.go:
```diff
@@ -348,3 +350,35 @@ func (r *resource153ToUnifiedResourceAdapter) CloneResource() ResourceWithLabels
+// marshaling the adapter itself).
+// If the resource153 is implemented by a protobuf type, we use
+// protojson.Marshal instead.
+func (r *resource153ToLegacyAdapterV2) MarshalJSON() ([]byte, error) {
```
should we add a unmarshalv2?

hugoShaka replied:
I don't think you can unmarshal into a resource 153 adapter, you unmarshal into the resource directly.

tigrato replied:
yeah, that's correct

codingllama commented on tool/tctl/common/collection_test.go:
```diff
@@ -431,3 +437,44 @@ func makeTestLabels(extraStaticLabels map[string]string) map[string]string {
 }
+
+// This test makes sure we marshal and unmarshal proto-based Resource153 properly.
+// We had a bug where types.Resource153 implemented by protobuf structs were not
```
> We had a bug 

Arguable whether it's a bug, given that Teleport does this at-large and changing the behavior is likely to cause bugs on itself.

hugoShaka replied:
Before the change, the following command chain failed:
```
tctl get autoupdate_config --format yaml > config.yaml
tctl create -f config.yaml
```

`tctl get`'s output not being accepted by `tctl create` will be considered a bug by the user.

codingllama replied:
It works as intended for various other types, so I think the blanket "it's a bug" is what tripped me. Some qualification is in order, otherwise why wouldn't we just fix the existing "Resource153ToLegacy" method?

hugoShaka replied:
> why wouldn't we just fix the existing "Resource153ToLegacy" method?

I absolutely want to do this, rending timestamps as objects with seconds/nanos is an inconsistency comparted to what we're doing for every other resource.I'm not doing this because I don't know who else calls `json.Marshal()` and I'm afraid someone might be relying on the wrong behaviour (e.g. we put a json containing the resource in audit logs, the elasticsearch mappings now expect the broken format).

codingllama commented on tool/tctl/common/collection_test.go:
```diff
@@ -431,3 +437,44 @@ func makeTestLabels(extraStaticLabels map[string]string) map[string]string {
+// We had a bug where types.Resource153 implemented by protobuf structs were not
+// marshaled properly (they should be marshaled using protojson). This test
+// checks we can do a round-trip with one of those proto-struct resource.
+func TestRoundTripProtoResource153(t *testing.T) {
```
The test name should mention Resource153ToLegacyV2, which it does rely on (unless I'm reading it wrong?).

It should also, ideally, compare the behavior with the regular Resource153ToLegacy which purposefully does not use protojon.

hugoShaka replied:
Done, added a failing test case.

codingllama commented on tool/tctl/common/collection_test.go:
```diff
@@ -431,3 +437,44 @@ func makeTestLabels(extraStaticLabels map[string]string) map[string]string {
+	collection := &autoUpdateConfigCollection{config: initial}
+	buf := &bytes.Buffer{}
+	require.NoError(t, writeYAML(collection, buf))
+	t.Log(buf.String())
```
nit: is this necessary?

```suggestion
```

codingllama commented on tool/tctl/common/collection_test.go:
```diff
@@ -431,3 +437,44 @@ func makeTestLabels(extraStaticLabels map[string]string) map[string]string {
+func TestRoundTripProtoResource153(t *testing.T) {
+	// Test setup: generate fixture.
+	initial, err := autoupdate.NewAutoUpdateConfig(&autoupdatev1pb.AutoUpdateConfigSpec{
+		Tools: nil,
```
nit:

```suggestion
```

codingllama reviewed (commented):
Drive-by as requested by Tim.

codingllama commented on tool/tctl/common/collection_test.go:
It looks like you have test failures? (TestCreateResources/autoupdate_agent_rollout / ./tool/tctl/common/resource_command_test.go)

hugoShaka replied:
Fixes the tests, they were doing json.Unmarshal.

hugoShaka commented on api/types/resource_153.go:
```diff
@@ -348,3 +350,35 @@ func (r *resource153ToUnifiedResourceAdapter) CloneResource() ResourceWithLabels
+//
+// Note that CheckAndSetDefaults is a noop for the returned resource and
+// SetSubKind is not implemented and panics on use.
+func Resource153ToLegacyV2(r Resource153) Resource {
```
> Should we update Resource153ToLegacy to mention Resource153ToLegacyV2?

I don't know in which context other resource153ToLegacy.MarshalJSON is used. This might break schemas and compatibility.

At least for resource manifests, yes we should always use protojson, else all the timestamps/durations and other google magical proto types are wrong.

> Could we choose a method name clearer than V2?

I'm open to any suggestion. Maybe `Resource153ToLegacyWithOptions` if we chose the options pattern?

> Alternatively, should we use the options pattern along with Resource153ToLegacy and keep it as a single method?

Sounds good to me, but this is a breaking change, this code is in `api`, and we're going to backport this at least up to v15. So this would break the semver-like versioning we have.

codingllama replied:
IMO adding variadic options is fine even for older api/ versions, but I'd also be OK with a non-breaking variant for release branches if you think it's worth it.

I do think that keeping it to a single method is worth doing for master, as we'll carry this forward for a long time.

espadolini replied:
IMO we should either have a separate method for "adapt this types.Resource153 that's also a proto.Message that should get marshaled with protojson into a types.Resource" or - the worst option of all but maybe the pragmatic option - we add an explicit list of concrete types that should be protojsoned in there. 😢

espadolini replied:
(outside of other pretty fringe options like setting a protobuf extension on the message and checking that at runtime or using that extension for a custom codegen that defines MarshalJSON/UnmarshalJSON as needed)

hugoShaka replied:
I changed the controversial V2 into `types.ProtoResource153ToLegacy` which explains more clearly the difference with the original one and doesn't have to check if the underlying resource implements proto.Message every time.

hugoShaka commented on api/types/resource_153.go:
> Ideally the codebase would always use protojson

Protojson only works on proto message, but `Resource153` can be both implemented by proto structs and manual go structs. We can't know if we should use `json` or `protojson` without a type assertion. Each time we will forget to check the type and overload `MarshalJSON` this will happen again.

codingllama replied:
I meant that the codebase should use protojson when marshaling protos, naturally.

I'm also generally +1 on explicit storage representations instead of simply marshaling whatever proto is in hand (which is typically a public surface/API proto). This is a different subject though.

hugoShaka commented on api/types/resource_153.go:
```diff
@@ -348,3 +350,35 @@ func (r *resource153ToUnifiedResourceAdapter) CloneResource() ResourceWithLabels
+func (r *resource153ToLegacyAdapterV2) MarshalJSON() ([]byte, error) {
+	if protoResource153, ok := r.inner.(proto.Message); ok {
+		return protojson.MarshalOptions{
+			UseProtoNames: true,
```
> it's a bit surprising that we are assuming it as the default

Anything user facing like resource manifests is using snake_case. I'm not sure how making this an option is an improvement as we would serialize using two different styles depending on the use-case/place. Enforcing a single format ensure we can't forget about setting the option and present a resource to the user with the wrong format.

codingllama replied:
Fair enough.

hugoShaka committed:
Update tool/tctl/common/collection_test.go

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

hugoShaka committed:
Update tool/tctl/common/collection_test.go

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

codingllama reviewed (commented):
@hugoShaka you requested a review but there are still plenty of open/unaddressed comments. I don't see significant changes in the PR either (apart from a couple of accepted suggestions). Maybe you forgot to push some commits?

hugoShaka committed:
Address feedback

- Change from Resource153AdapterV2 to ProtoResource153Adapter
- fix test failures and unmarshal proto resources properly
- add a failing round-trip proto 153 test case
- bonus: fix the table tesst reosurce create that did not support
  running a single row

hugoShaka commented on tool/tctl/common/resource_command_test.go:
```diff
@@ -1371,17 +1372,29 @@ func TestCreateResources(t *testing.T) {
 	process := testenv.MakeTestServer(t, testenv.WithLogger(utils.NewSlogLoggerForTests()))
 	rootClient := testenv.MakeDefaultAuthClient(t, process)
 
+	// tctlGetAllValidations allows tests to register post-test validations to validate
+	// that their resource is present in "ttcl get all" output.
+	// This allows running test rows instead of the whole test table.
+	var tctlGetAllValidations []func(t *testing.T, out string)
+
```
Sorry for the noisy changes in the table test but I got fed up by the test failing if I only ran a single row. Now the individual rows can be ran.

codingllama commented on api/types/resource_153.go:
```diff
@@ -348,3 +354,38 @@ func (r *resource153ToUnifiedResourceAdapter) CloneResource() ResourceWithLabels
+
+// MarshalJSON adds support for marshaling the wrapped resource (instead of
+// marshaling the adapter itself).
+// If the resource153 is implemented by a protobuf type, we use
+// protojson.Marshal instead.
```
```suggestion
```

codingllama commented on tool/tctl/common/resource_command_test.go:
```diff
@@ -1371,17 +1372,29 @@ func TestCreateResources(t *testing.T) {
 	rootClient := testenv.MakeDefaultAuthClient(t, process)
 
+	// tctlGetAllValidations allows tests to register post-test validations to validate
+	// that their resource is present in "ttcl get all" output.
```
```suggestion
	// that their resource is present in "tctl get all" output.
```

codingllama reviewed (approved):
Thanks!

hugoShaka committed:
Apply suggestions from code review

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

hugoShaka committed:
lint
